### PR TITLE
[SPARK-26089][CORE] Handle corruption in large shuffle blocks

### PR DESCRIPTION
**Pull request by ankuriitg**
## What changes were proposed in this pull request?

SPARK-4105 added corruption detection in shuffle blocks but that was limited to blocks which are
smaller than maxBytesInFlight/3. This commit adds upon that by adding corruption check for large
blocks. There are two changes/improvements that are made in this commit:

1. Large blocks are checked upto maxBytesInFlight/3 size in a similar way as smaller blocks, so if a
large block is corrupt in the starting, that block will be re-fetched and if that also fails,
FetchFailureException will be thrown.
2. If large blocks are corrupt after size maxBytesInFlight/3, then any IOException thrown while
reading the stream will be converted to FetchFailureException.  This is slightly more aggressive
than was originally intended but since the consumer of the stream may have already read some records and processed them, we can't just re-fetch the block, we need to fail the whole task. Additionally, we also thought about maybe adding a new type of TaskEndReason, which would re-try the task couple of times before failing the previous stage, but given the complexity involved in that solution we decided to not proceed in that direction.

Thanks to @squito for direction and support.

## How was this patch tested?

Changed the junit test for big blocks to check for corruption.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -332,6 +332,41 @@ private[spark] object Utils extends Logging {
...
+          out.write(buf, 0, n)
+          count += n
+        }
+      }
+      count < maxSize
+    } {
+      if (closeStreams) {
+        try {
+          if (count < maxSize) {
+            in.close()
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: maxSize might be exactly the same length as the stream, in which case you'll read to the end, but not close it.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -332,6 +332,41 @@ private[spark] object Utils extends Logging {
     }
   }
 
+  /**
+   * Copy all data from an InputStream to an OutputStream upto maxSize and
+   * closes the input stream if closeStreams is true and all data is read.
+   */
+  def copyStreamUpto(
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: super nit: capital T, UpTo

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 3]
```diff
@@ -453,14 +454,21 @@ final class ShuffleBlockFetcherIterator(
             input = streamWrapper(blockId, in)
             // Only copy the stream if it's wrapped by compression or encryption, also the size of
             // block is small (the decompressed block is smaller than maxBytesInFlight)
-            if (detectCorrupt && !input.eq(in) && size < maxBytesInFlight / 3) {
+            if (detectCorrupt && !input.eq(in)) {
               isStreamCopied = true
+              streamCompressedOrEncrypted = true
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I wonder if we should do the inputStream checks even if `detectCorrupt` is false, and perhaps even change that default.  Or add another flag just for doing this memory copy, as its more expensive.

Also I think the comments around here need to be updated to explain what is happening now.

**Review by squito — COMMENTED** [review]
just a partial review so far

there should be a test that a large block is read correctly, with the partial buffering & stream concatenation.  I think most unit tests will not end up creating a large block.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 1 above)
Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: Right, I left that edge case, as it will be handled by the concatenated stream later

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 3 above)
Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: Sure, I can change it to always detect corruption. Which default did you mean though, the default value of `detectCorrupt` is true.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 4]
```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
...
               val out = new ChunkedByteBufferOutputStream(64 * 1024, ByteBuffer.allocate)
-              // Decompress the whole block at once to detect any corruption, which could increase
-              // the memory usage tne potential increase the chance of OOM.
+              // Decompress the block upto maxBytesInFlight/3 at once to detect any corruption which
+              // could increase the memory usage and potentially increase the chance of OOM.
               // TODO: manage the memory used here, and spill it into disk in case of OOM.
-              Utils.copyStream(input, out, closeStreams = true)
-              input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              isStreamCopied = Utils.copyStreamUpTo(
+                input, out, maxBytesInFlight / 3, closeStreams = true)
```
Comment (rezasafi) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: What is the reasoning behind considering one-third of the block, why for example not one-fourth?

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 5]
```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
...
+              if (isStreamCopied) {
+                input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              } else {
```
Comment (rezasafi) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: What is the behavior here if the stream wasn't copied because of the corruption? It seems that this else ignores the corruption in one-third and will try to see if the corruption is happening  in the remaining two-third which seems not necessary.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 6]
```diff
@@ -490,7 +499,9 @@ final class ShuffleBlockFetcherIterator(
     }
 
     currentResult = result.asInstanceOf[SuccessFetchResult]
-    (currentResult.blockId, new BufferReleasingInputStream(input, this))
+    (currentResult.blockId,
```
Comment (rezasafi) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: What is the purpose of this new updated return? currentResult.blockId is already returning in tuple.

**Review by rezasafi — COMMENTED** [review]
Thanks for the work here. This is just an initial review to better understand your approach. I think It will be great to ping the original authors of SPARK-4105 and related changes to get their feedback as well.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 5 above)
Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I am guessing you did not look into Utils.copyStreamUpTo(). isStreamCopied is true when the size of the stream is less than maxBytesInFlight/3 and that stream was copied. If it is longer, then it will go to the else block. If stream wasn't copied because of corruption, it will be handled in catch block.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 4 above)
Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: That limit was already defined. I did not change that limit. The purpose was to limit the amount of memory used but I am not sure how we came up with that exact threshold.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 6 above)
Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: It is still returning the same tuple. I am just passing more arguments to the constructor of BufferReleasingInputStream

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 5 above)
Comment (rezasafi) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: A little confusing if this if only check the size but its condition only says isStreamCopied.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 4 above)
Comment (rezasafi) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I can't find where it is defined. I thought your change is proposing that. Sorry if I missed that

Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: Look for this line on left side of the files changed: `if (detectCorrupt && !input.eq(in) && size < maxBytesInFlight / 3) {`

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 5 above)
Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: Please look at `Utils.copyStreamUpTo()`

**Comment by ankuriitg** [comment]
cc @davies since you worked on SPARK-4105

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 7]
```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
...
-              Utils.copyStream(input, out, closeStreams = true)
-              input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              isStreamCopied = Utils.copyStreamUpTo(
+                input, out, maxBytesInFlight / 3, closeStreams = true)
+              if (isStreamCopied) {
+                input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              } else {
+                input = new SequenceInputStream(
+                  out.toChunkedByteBuffer.toInputStream(dispose = true), input)
+              }
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I think it would be better if you refactored this so `copyStreamUpTo` includes this part as well, and just returns an InputStream.  The main reason is for testing -- I worry about edge cases here.   If this is factored out, it would be much easier to test that the resulting input stream is OK if eg. there is an empty input, or just up to the limit, or over the limit by X, and make sure streams get closed correctly etc.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 8]
```diff
@@ -332,6 +332,41 @@ private[spark] object Utils extends Logging {
...
+   * closes the input stream if closeStreams is true and all data is read.
+   */
+  def copyStreamUpTo(
+      in: InputStream,
+      out: OutputStream,
+      maxSize: Long,
+      closeStreams: Boolean = false): Boolean = {
+    var count = 0L
+    tryWithSafeFinally {
+      val bufSize = 8192
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: max(8192, maxSize)

**Review by squito — COMMENTED** [review]
>> I wonder if we should do the inputStream checks even if detectCorrupt is false, and perhaps even change that default. Or add another flag just for doing this memory copy, as its more expensive.

> Which default did you mean though, the default value of detectCorrupt is true.

Sorry I didn't explain that very well.  What I was wondering is we should change the behavior to

* max the copy-into-memory check turned off by default
* but *always* do the on-the-fly check, even if detectCorrupt is false.  You could add another option for this also, and have it default true, if really want to allow users to turn off all detection.

I'm saying this because I think that I'd want to run my spark apps with the new default I'm suggesting -- don't do the copy into memory, but do the on-the-fly check all the time, of the entire block.

**Comment by ankuriitg** [comment]
retest this please

**Comment by squito** [comment]
real build error:

```
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/internal/config/package.scala:931: SHUFFLE_DETECT_CORRUPT is already defined as value SHUFFLE_DETECT_CORRUPT
[error]   private[spark] val SHUFFLE_DETECT_CORRUPT =
[error]                      ^
[error] one error found
```

**Comment by ankuriitg** [comment]
> real build error:
> 
> ```
> [error] /home/jenkins/workspace/SparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/internal/config/package.scala:931: SHUFFLE_DETECT_CORRUPT is already defined as value SHUFFLE_DETECT_CORRUPT
> [error]   private[spark] val SHUFFLE_DETECT_CORRUPT =
> [error]                      ^
> [error] one error found
> ```

Thanks, I realized it after this morning's build failure. The previous test failure confused me and I did not see the build failure before that.

**Review by squito — COMMENTED** [review]
just some test updates.

@tgravescs how do you feel about my suggested updates to the defaults?

**Comment by tgravescs** [comment]
I think you are talking about:

>  but always do the on-the-fly check, even if detectCorrupt is false. You could add another option for this also, and have it default true, if really want to allow users to turn off all detection.

so the only place detectCorrupt is used after this change is  passed into BufferReleasingInputStream 
 for the read, and its default is true.  If we want to make the exception handling there always present, are you saying just remove the detectCorrupt config or are you suggesting something with the streamCompressedOrEncrypted?

**Comment by squito** [comment]
sorry my earlier comment wasn't clear -- ankur was confused too but I think this version of the patch implements what I think the right option is:

* by default, do not copy into memory at all, but do on-the-fly detection of corruption.  This alway results in a fetch-failure
* optionally, go back to the old behavior, to copy into memory if the block is small.  Then you'd first retry pulling the data, and throw a fetch failure after that.  Turned on with "spark.shuffle.detectCorrupt.useExtraMemory=true"
* optionally, never do any check, even on the fly (using the old config, "spark.shuffle.detectCorrupt=false")

**Comment by tgravescs** [comment]
Ok, That makes sense to me and agree with you.

File: core/src/test/scala/org/apache/spark/util/UtilsSuite.scala [hunk 9]
```diff
@@ -211,6 +213,49 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
...
   }
 
+  test("copyStreamUpTo") {
+    // input array initialization
+    val bytes = Array.ofDim[Byte](1200)
+    Random.nextBytes(bytes)
+
+    val limit = 1000
+    // testing for inputLength less than, equal to and greater than limit
+    List(900, 1000, 1100).foreach { inputLength =>
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/util/UtilsSuite.scala]: sorry I'm being very nitpicky here, but I really want to make sure we're testing the exact edge conditions when the limit is -2, -1, 0, +1, +2 of the full length.  Its OK if you don't free the buffer till you are past it, but it should happen shortly after the limit.

File: core/src/test/scala/org/apache/spark/util/UtilsSuite.scala [hunk 10]
```diff
@@ -211,6 +213,49 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
...
+        assert(byteBufferInputStream.chunkedByteBuffer === null)
+      } finally {
+        IOUtils.closeQuietly(mergedStream)
+        IOUtils.closeQuietly(in)
+      }
+    }
+  }
+
+  private def getFieldValue(obj: AnyRef, fieldName: String): Any = {
+    val mirror = universe.runtimeMirror(obj.getClass().getClassLoader())
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/util/UtilsSuite.scala]: Use java reflection instead of scala reflection here -- its known to more developers, has fewer strange gotchas, simpler and is sufficient for this.

File: core/src/test/scala/org/apache/spark/util/UtilsSuite.scala [hunk 11]
```diff
@@ -211,6 +213,49 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
...
+    val bytes = Array.ofDim[Byte](1200)
+    Random.nextBytes(bytes)
+
+    val limit = 1000
+    // testing for inputLength less than, equal to and greater than limit
+    List(900, 1000, 1100).foreach { inputLength =>
+      val in = new ByteArrayInputStream(bytes.take(inputLength))
+      val (fullyCopied: Boolean, mergedStream: InputStream) = Utils.copyStreamUpTo(in, limit, true)
+      try {
+        val byteBufferInputStream = if (mergedStream.isInstanceOf[ChunkedByteBufferInputStream]) {
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/util/UtilsSuite.scala]: Add some comments here, like "Get a handle on the buffered data, to make sure the memory gets freed once we read past the end of it.  Need to use refelction to get handle on inner structures for this check"

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 12]
```diff
@@ -461,10 +485,68 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
       Int.MaxValue,
       Int.MaxValue,
       true,
+      true,
+      taskContext.taskMetrics.createTempShuffleReadMetrics())
+    // Only one block should be returned which has corruption after maxBytesInFlight/3
+    val (id, st) = iterator.next()
+    assert(id === ShuffleBlockId(0, 1, 0))
+    intercept[FetchFailedException] { iterator.next(); iterator.next() }
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala]: I'm a little confused here -- why do you need to call iterator.next() twice?

**Comment by squito** [comment]
Jenkins, retest this please

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 13]
```diff
@@ -335,6 +336,47 @@ private[spark] object Utils extends Logging {
...
   }
 
+  /**
+   * Copy all data from an InputStream to an OutputStream upto maxSize and
+   * closes the input stream if closeStreams is true and all data is read.
+   */
+  def copyStreamUpTo(
+      in: InputStream,
+      maxSize: Long,
+      closeStreams: Boolean = false): (Boolean, InputStream) = {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: I don't think you need to expose `closeStreams`, just have it always be true.  At least I can't think of a reason why it should default to `false` (`copyStream` is used more places ...)

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 14]
```diff
@@ -571,7 +582,8 @@ final class ShuffleBlockFetcherIterator(
     }
   }
 
-  private def throwFetchFailedException(blockId: BlockId, address: BlockManagerId, e: Throwable) = {
+  private[storage] def throwFetchFailedException(
+      blockId: BlockId, address: BlockManagerId, e: Throwable) = {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: nit: if the method declaration is multi-line, each arg on its own line.  And this method should have had a return type in the first place (probably my fault, oops)

```scala
private[storage] def throwFetchFailedException(
    blockId: BlockId,
    address: BlockManagerId,
    e: Throwable): Unit = {
```

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 15]
```diff
@@ -583,15 +595,28 @@ final class ShuffleBlockFetcherIterator(
...
  */
 private class BufferReleasingInputStream(
-    private val delegate: InputStream,
-    private val iterator: ShuffleBlockFetcherIterator)
+    // This is visible for testing
+    private[storage] val delegate: InputStream,
+    private val iterator: ShuffleBlockFetcherIterator,
+    private val blockId: BlockId,
+    private val address: BlockManagerId,
+    private val streamCompressedOrEncrypted: Boolean)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I would rename `streamCompressedOrEncrypted` to `detectCorruption` or something like that, as the condition is a bit more complex now (you also check the `detectCorrupt` config when passing this in)

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 16]
```diff
@@ -335,6 +336,44 @@ private[spark] object Utils extends Logging {
...
+  /**
+   * Copy all data from an InputStream to an OutputStream upto maxSize and
+   * close the input stream if all data is read.
+   * @return A tuple of boolean, which is whether the stream was fully copied, and an InputStream,
+   *         which is a combined stream of read data and any remaining data
+   */
+  def copyStreamUpTo(in: InputStream, maxSize: Long): (Boolean, InputStream) = {
+    var count = 0L
+    val out = new ChunkedByteBufferOutputStream(64 * 1024, ByteBuffer.allocate)
+    val streamCopied = tryWithSafeFinally {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: nit: I think a better name for `streamCopied` would be `fullyCopied` ... everytime I come back to this code I always get a bit confused that maybe no copying is happening in some situations.

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 17]
```diff
@@ -425,28 +450,97 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
...
+      Int.MaxValue,
+      Int.MaxValue,
+      Int.MaxValue,
+      true,
+      true,
+      taskContext.taskMetrics.createTempShuffleReadMetrics())
+
+    // Only one block should be returned which has corruption after maxBytesInFlight/3 because the
+    // other block will detect corruption on first fetch, and then get added to the queue again for
+    // a retry
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala]: I'd reword this -- you get one block because you call next() once.  You really want to explain why you get a certain block back.

```We'll get back the block which has corruption after maxBytesInFlight/3 because ...```

(assuming I understand this correctly...)

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 18]
```diff
@@ -425,28 +450,97 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
...
+    val (id, st) = iterator.next()
+    assert(id === shuffleBlockId2)
+
+    // The other block will throw a FetchFailedException
+    intercept[FetchFailedException] {
+      iterator.next()
+    }
+
+    // Following will succeed as it reads part of the stream which is not corrupt. This will read
+    // maxBytesInFlight/3 bytes from first stream and remaining from the second stream
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala]: ... will read maxBytesInFlight/3 bytes from the portion copied into memory, and the remaining from the underlying stream.

File: core/src/test/scala/org/apache/spark/util/UtilsSuite.scala [hunk 19]
```diff
@@ -211,6 +213,55 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
...
   }
 
+  test("copyStreamUpTo") {
+    // input array initialization
+    val bytes = Array.ofDim[Byte](1200)
+    Random.nextBytes(bytes)
+
+    val limit = 1000
+    // testing for inputLength less than, equal to and greater than limit
+    List(998, 999, 1000, 1001, 1002).foreach { inputLength =>
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/util/UtilsSuite.scala]: `(limit - 2 to limit + 2).foreach`

**Review by squito — APPROVED** [review]
a few very minor style things, otherwise lgtm

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 20]
```diff
@@ -466,16 +469,19 @@ final class ShuffleBlockFetcherIterator(
...
+            // maxBytesInFlight/3. If stream is longer, then corruption will be caught while reading
+            // the stream.
+            streamCompressedOrEncrypted = !input.eq(in)
+            if (streamCompressedOrEncrypted && detectCorruptUseExtraMemory) {
               isStreamCopied = true
-              val out = new ChunkedByteBufferOutputStream(64 * 1024, ByteBuffer.allocate)
-              // Decompress the whole block at once to detect any corruption, which could increase
-              // the memory usage tne potential increase the chance of OOM.
+              // Decompress the block upto maxBytesInFlight/3 at once to detect any corruption which
+              // could increase the memory usage and potentially increase the chance of OOM.
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: this comment & the one just above the `if` are redundant and a little bit wrong -- I think you only need one comment (don't care whether its above or below the if) and should be something more like:

We optionally decompress the first maxBytesInFlight/3 bytes into memory, to check for corruption in that portion of the data.  But even if that configuration is off, or if the corruption is later, we'll still try to detect the corruption later in the stream.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 21]
```diff
@@ -466,16 +469,19 @@ final class ShuffleBlockFetcherIterator(
...
               // TODO: manage the memory used here, and spill it into disk in case of OOM.
-              Utils.copyStream(input, out, closeStreams = true)
-              input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              val (fullyCopied: Boolean, mergedStream: InputStream) = Utils.copyStreamUpTo(
+                input, maxBytesInFlight / 3)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I'm trying to understand why the

```scala
finally {
            // TODO: release the buf here to free memory earlier
            if (isStreamCopied) {
              in.close()
            }
```

is needed down below.  To be honest, I don't think it was needed in the old code.  The old `Utils.copyStream` was always called with `closeStreams=true`, and that would always close the input in a `finally` itself:

https://github.com/apache/spark/blob/d9978fb4e4d4de3a320b012373c18bd278462780/core/src/main/scala/org/apache/spark/util/Utils.scala#L302-L307

https://github.com/apache/spark/blob/d9978fb4e4d4de3a320b012373c18bd278462780/core/src/main/scala/org/apache/spark/util/Utils.scala#L330-L332

It doesn't hurt, but it also makes things unnecessarily confusing.  If you didn't need to do that `in.close()` below, you woudln't need to track `isStreamCopied`, and wouldn't need to even return `fullyCopied` from `Utils.copyStreamUpTo`.  That's really the part of this which is bugging me -- something seems off that we need to know whether or not the stream is fully copied, seems like it shouldn't matter.  If it does matter, aren't we getting something wrong in the case where the stream is *exactly* maxBytesInFlight / 3, but we haven't realized its fully copied because we haven't read past the end yet?

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 22]
```diff
@@ -466,16 +469,19 @@ final class ShuffleBlockFetcherIterator(
...
+              isStreamCopied = fullyCopied
+              input = mergedStream
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: not related to your changes, but while you're touching this file, can you add 2 more spaces of indentation to `|| corruptedBlocks.contains(blockId)) {` a couple of lines below this?

**Review by squito — COMMENTED** [review]
Thanks for the update @ankuriitg .  sorry I am being very particular about this, its just a really core piece.  I have a couple of minor updates to comments -- but really I was doing one final pass and I got pretty confused about the existing `in.close()`.  I'd like us to clean that up now, while we're looking at this, if possible.   Left a longer comment inline, please check my reasoning.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 21 above)
Comment (ankuriitg) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I think in.close() is needed if there is an exception while creating a wrapped stream. The first time I saw `isStreamCopied`, I was also confused and by looking at it more closely now, I realize that it is not doing what it is supposed to do.

I have removed `isStreamCopied` and instead used another condition to close the stream. Please check and let me know if it makes sense.

**Comment by squito** [comment]
merged to master, thanks @ankuriitg !
